hostap: Security improvements

MaochenWang1 opened the pull request:
hostap: Support getting Wi-Fi connection parameters recently used
hostap: Support flushing PMKSA cache entries
hostap: add WPA2 EAP_TLS support

Implements https://github.com/zephyrproject-rtos/zephyr/issues/54213 partially (only EAP-TLS).

MaochenWang1 committed:
hostap: Support getting Wi-Fi connection parameters recently used

Support saving and getting Wi-Fi connection parameters recently used.

Signed-off-by: Maochen Wang <maochen.wang@nxp.com>

MaochenWang1 committed:
hostap: Support flushing PMKSA cache entries

Support flushing PMKSA cache entries in the reconnection
failed case of WPA3 SAE.

Signed-off-by: Maochen Wang <maochen.wang@nxp.com>

MaochenWang1 committed:
wifi: Add WPA2 EAP-TLS support

Add basic WPA2 EAP-TLS support.
Also, add test infrasturcture esp. the certification handling,
non-certificate credentials are take as runtime input and certificated
are build time input for testing.

A real application can set certificates at runtime too.

Signed-off-by: Maochen Wang <maochen.wang@nxp.com>
Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -24,6 +24,9 @@
 extern struct wpa_global *global;
 
+/* save the last wifi connection parameters */
+static struct wifi_connect_req_params last_wifi_conn_params;
```
can you please add to the commit log, how/where is this used?

MaochenWang1 replied:
it's used in NXP wifi code, user can decide to get the recently used wifi connection parameters

krish2718 replied:
it would be better to add this to Wi-Fi shell so that we can test/see the usage of the API.

MaochenWang1 replied:
Hi @krish2718 , I carefully think about this, and I find it's difficult to add this to wifi shell, as structure of wifi_connect_req_params has too many paramters ( and continue growing), and difficult to parse and print all of them.
if must add the caller, can I only add in wifi_mgmt.c, not the wifi_shell.c?

krish2718 commented on modules/hostap/src/supp_api.h:
```diff
@@ -73,6 +73,14 @@ int supplicant_scan(const struct device *dev, struct wifi_scan_params *params,
+ *
+ * @return 0 if ok, < 0 if error
+ */
+int supplicant_pmksa_flush(const struct device *dev);
```
Same, please add how/where is this used?

MaochenWang1 replied:
It's used in NXP wifi code, and user can decide to use this API to Flush PMKSA cache entries

krish2718 replied:
it would be better to add this to Wi-Fi shell so that we can test/see the usage of the API.

MaochenWang1 replied:
Hi @krish2718 can I only add this into wifi_mgmt.c?

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -402,6 +402,14 @@ struct wifi_connect_req_params {
 	/** Connect timeout in seconds, SYS_FOREVER_MS for no timeout */
 	int timeout;
+	/** anonymous identity */
+	const uint8_t *aid;
```
`anon_id` is better to disambiguate with `Association ID`

MaochenWang1 replied:
I can change it

krish2718 commented on modules/hostap/src/supp_api.h:
```diff
@@ -19,6 +19,17 @@
 #define CHAN_NUM_LEN 6 /* for space-separated channel numbers string */
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE
+struct wpa_supp_enterprise_file {
```
how are these taken as input from the user? Is that support coming in upcoming PRs?

krish2718 replied:
it would be better to add this to Wi-Fi shell so that we can test/see the usage of the API.

MaochenWang1 replied:
Vendor can decide when to set the enterprise file.
supplicant_add_enterprise_file is called in NXP wifi code.

MaochenWang1 replied:
Hi @krish2718 , can I only add the caller of supplicant_add_enterprise_file into wifi_mgmt.c, not the wifi_shell.c?
as I find input the shell to add enterprise file is weird.

krish2718 replied:
> as I find input the shell to add enterprise file is weird.

Can you please elaborate? `wifi pmksa_flush` command should be fine right?

MaochenWang1 replied:
okay, wifi pmksa_flush, I will add to wifi shell

MaochenWang1 commented:
Hi @krish2718 @jukkar @rlubos please review again, thanks

krish2718 commented on modules/hostap/src/supp_api.c:
```diff
@@ -905,6 +910,27 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
+		goto out;
+	}
+
+	params = &last_wifi_conn_params;
```
this would still be a shallow-copy, so, may be use memcpy

MaochenWang1 replied:
updated

krish2718 commented on include/zephyr/net/wifi.h:
```diff
@@ -118,6 +118,10 @@ const char *wifi_band_txt(enum wifi_frequency_bands band);
 #define WIFI_SAE_PSWD_MAX_LEN 128
 /** MAC address length */
 #define WIFI_MAC_ADDR_LEN 6
+/** Max identity length */
+#define WIFI_IDENTITY_MAX_LEN 64
+/** Max password length */
+#define WIFI_PSWD_MAX_LEN 128
```
may be add a prefix to indicate that this is for for Enterprise?

MaochenWang1 replied:
updated

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -95,6 +95,8 @@ enum net_request_wifi_cmd {
 	/** Flush PMKSA cache entries */
 	NET_REQUEST_WIFI_CMD_PMKSA_FLUSH,
+	/** Set enterprise mode file */
+	NET_REQUEST_WIFI_CMD_ENTERPRISE_FILE,
```
`FILE` isn't the right term here, `ENTERPRISE_CREDS` or `CERT_INFO`?

MaochenWang1 replied:
changed to ENTERPRISE_CREDS

MaochenWang1 committed:
hostap: rename WIFI_SECURITY_TYPE_EAP

Keep WIFI_SECURITY_TYPE_EAP, and define WIFI_SECURITY_TYPE_EAP_TLS
same value as WIFI_SECURITY_TYPE_EAP to make it backwards
compatible. Ready to support more EAP type in the comming PR.

Signed-off-by: Maochen Wang <maochen.wang@nxp.com>

MaochenWang1 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2021,12 +2042,14 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		  "[-m, --bssid]: MAC address of the AP (BSSID).\n"
 		  "[-t, --timeout]: Timeout for the connection attempt (in seconds).\n"
+		  "[-a, --anon-id]: Anonymous identity for enterprise mode.\n"
+		  "[-K, --key-passwd]: Private key passwd for enterprise mode.\n"
```
How to add certs to wifi at build time? I think you can call the API to set the enterprise creds in your driver init phase.

MaochenWang1 replied:
maybe this can be done in the future as enhancement?

krish2718 replied:
As discussed last in the Wi-Fi telco, you can put the raw cert in a file and can include it in a C array. See below links
https://github.com/nrfconnect/sdk-nrf/tree/v2.6.1/samples/net/aws_iot/certs
https://github.com/nrfconnect/sdk-nrf/blob/main/samples/net/https_client/src/main.c#L46

> maybe this can be done in the future as enhancement?
I think you can call the API to set the enterprise creds in your driver init phase.

It would be good to get this in this PR itself, it shouldn't be too hard, this way anyone an test those APIs using shell rather than writing code to invoke the API.

MaochenWang1 replied:
@krish2718 I can put the cert/key as C array in header files. but where should I put the header file? and what is the default value of the cert/key, just the one we currently use?

krish2718 replied:
> but where should I put the header file

It's a bit tricky, but I would just put them in `subsys/net/l2/wifi/test_certs` and include those in `wifi_shell.c` if Enterprise is enabled and use in `set_enterprise_creds`. @jukkar WDYT? The ideas is to use `shell` as primary test method.

> and what is the default value of the cert/key, just the one we currently use?

Sure, or you generate a new test ones.

MaochenWang1 replied:
Hi @krish2718  @jukkar , updated as your comment, please review again

krish2718 replied:
Thanks, this is looking good, but it would be good to commit certs directly and then have a script or some way to generate the headers, this makes it easier to update these certs. WDYT?

I used this script:

```
 cat ~/crt_to_header.py 
#!/usr/bin/env python3
import sys

cer_file = sys.argv[1]
header_file = sys.argv[2]
macro_name = "CA_CERTIFICATE"

key = open(cer_file, "r")
c_file = open(header_file, "w")
c_file.write("#define " + macro_name + " \\\n")
for line in key:
    line = line.replace("\n", "")
    c_line = "\"" + line + "\\n\" \\\n"
    c_file.write(c_line)
print('Certificate converted to C header file in:', header_file)
key.close()
c_file.close()
```

This is based on https://academy.nordicsemi.com/courses/cellular-iot-fundamentals/lessons/lesson-4-cellular-fundamentals/topic/lesson-4-exercise-2/ (#3.2). No changes to source code, just replace cert and then run the script.

MaochenWang1 replied:
Hi @krish2718 , if directly using certs file to replace the header files, then the build can not pass:
![image](https://github.com/user-attachments/assets/bf80f944-8d6c-43c3-85af-b2c5fde1e990)
as we need to use the C array in header files. why not let the user to use their way to generate the header files? 
we use this way:

Convert CA format form PEM to DER.
openssl x509 -outform der -in cas.pem -out ca.der
openssl x509 -outform der -in wifiuser.pem -out client.der

Convert RSA Key format to DER.
openssl rsa -in wifiuser.key -outform DER -out client_key.der

Convert certs to header file:
Convert .der to .h: xxd -i ca.der > ca.h

krish2718 replied:
I am suggesting to keep both the original cert file and the generated header, not just the cert file. And also commit the script, so, user can just update cert file run and generate header.

MaochenWang1 replied:
can I add the pem file and header file, and a readme.txt to tell user how to convert pem to header file?
Convert CA format form PEM to DER.
openssl x509 -outform der -in cas.pem -out ca.der
openssl x509 -outform der -in wifiuser.pem -out client.der

Convert RSA Key format to DER.
openssl rsa -in wifiuser.key -outform DER -out client_key.der

Convert certs to header file:
Convert .der to .h: xxd -i ca.der > ca.h

MaochenWang1 replied:
Hi @krish2718  please review again

MaochenWang1 commented on modules/hostap/src/supp_api.h:
```diff
@@ -19,6 +19,17 @@
 #define CHAN_NUM_LEN 6 /* for space-separated channel numbers string */
 
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE
+struct wpa_supp_enterprise_file {
```
change to enterprise_creds

MaochenWang1 commented on include/zephyr/net/wifi.h:
```diff
@@ -48,14 +48,14 @@ enum wifi_security_type {
 	/** GB 15629.11-2003 WAPI security. */
 	WIFI_SECURITY_TYPE_WAPI,
-	/** EAP security - Enterprise. */
-	WIFI_SECURITY_TYPE_EAP,
```
done

MaochenWang1 commented:
Hi @krish2718 @jukkar @rlubos please review again

krish2718 commented on subsys/net/l2/wifi/test_certs/README.txt:
```diff
@@ -0,0 +1,14 @@
+****************************************************************
+1. Convert CA format from PEM to DER.
+openssl x509 -outform der -in cas.pem -out ca.der
+openssl x509 -outform der -in wifiuser.pem -out client.der
+
+2. Convert RSA Key format to DER.
+openssl rsa -in wifiuser.key -outform DER -out client_key.der
+
+3. Convert .der to header file:
+xxd -i ca.der > ca.h
+
+4. Replace corresponding cert header files at below path:
+subsys/net/l2/wifi/test_certs/
+#please make sure the parameter names are same.
```
We should move this to `doc/connectivity/networking/api/wifi.rst` under a sub-section for `Wi-Fi shell enterprise testing: Certification provisioning`

```suggestion
.. _wifi_shell_test_certs:

X.509 Certificate header generation
###################################

Wi-Fi enterprise security requires use of X.509 certificates, a test certificates
in PEM format are committed to the repo and the below steps to be followed to convert
them to a C header file that can be included in Wi-Fi shell module.

Convert CA format from PEM to DER
=================================

To convert the CA format from PEM to DER, use the following command:

.. code-block:: none

    openssl x509 -outform der -in cas.pem -out ca.der

Convert RSA Key format to DER
=============================

To convert the RSA Key format to DER, use the following command:

.. code-block:: none

    openssl rsa -in wifiuser.key -outform DER -out client_key.der

Convert .der to header file
===========================

To convert the .der file to a header file, use the following command:

.. code-block:: none

    xxd -i ca.der > ca.h

Replace corresponding cert header files
=======================================

Replace the corresponding cert header files at the following path:

```
subsys/net/l2/wifi/test_certs/
```

.. note:: Please ensure that the parameter names are the same.
```

MaochenWang1 replied:
Hi @krish2718 , updated, please review again

krish2718 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -28,3 +28,36 @@ API Reference
+Convert CA format from PEM to DER
+=================================
+To convert the CA format from PEM to DER, use the following command:
+.. code-block:: none
+openssl x509 -outform der -in cas.pem -out ca.der
```
This needs a specific format:
```suggestion
.. code-block:: none

     openssl x509 -outform der -in cas.pem -out ca.der
```
See for an example https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/main/samples/net/wifi/README.rst

krish2718 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -28,3 +28,36 @@ API Reference
+Replace the corresponding cert header files at the following path:
+subsys/net/l2/wifi/test_certs/
+
+.. note:: Please ensure that the parameter names are the same.
```
Format to be fixed.

```suggestion
.. note:: 

   Please ensure that the parameter names in the header are unchanged, only the payload is updated.
```

MaochenWang1 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -28,3 +28,36 @@ API Reference
 
 .. doxygengroup:: wifi_mgmt
+
+.. _wifi_shell_test_certs:
```
Hi @krish2718 how to change this?
.. _wifi_shell_test_certs:

MaochenWang1 replied:
Hi @krish2718 
please review again

krish2718 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -28,3 +28,40 @@ API Reference
 
 .. doxygengroup:: wifi_mgmt
+
+X.509 Certificate header generation
```
Looking at https://builds.zephyrproject.io/zephyr/pr/74847/docs/connectivity/networking/api/wifi.html

this should be before the doxygen docs i.e., before `API reference`

krish2718 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -28,3 +28,40 @@ API Reference
 
 .. doxygengroup:: wifi_mgmt
+
+X.509 Certificate header generation
```
```suggestion
Wi-Fi Enterprise test: X.509 Certificate header generation
```

MaochenWang1 replied:
Hi @krish2718 
please review again

MaochenWang1 replied:
Hi @krish2718 @jukkar  please help review again

MaochenWang1 commented:
Hi @krish2718 @jukkarplease review again, thanks

MaochenWang1 commented:
Hi @krish2718 @jukkar @rlubos  please review again, thanks

krish2718 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -24,6 +24,43 @@ Currently, two types of Wi-Fi drivers are supported:
+Convert CA format from PEM to DER
+=======================================
+To convert the CA format from PEM to DER, use the following command:
+.. code-block:: none
```
This isn't being rendered properly, needs a newline before as well.

MaochenWang1 replied:
Hi @krish2718 thanks, please review again

krish2718 replied:
FYI, when you do a force push after resolving comments, reviewers get an email, and you can still trigger a specific email for a reviewer by clicking on `refresh` button as below.

![image](https://github.com/user-attachments/assets/425bdc8e-0ec6-4796-9d31-dde95ec19508)

MaochenWang1 replied:
thanks for your info

krish2718 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -24,6 +24,46 @@ Currently, two types of Wi-Fi drivers are supported:
 * Native L2 Ethernet drivers
 
+Wi-Fi Enterprise test: X.509 Certificate header generation
+##########################################################
```
Sorry, I gave you wrong heading format. This should be as per https://docs.zephyrproject.org/latest/contribute/documentation/guidelines.html#headings 

```suggestion
**************************************************
```

krish2718 commented on doc/connectivity/networking/api/wifi.rst:
```diff
@@ -24,6 +24,46 @@ Currently, two types of Wi-Fi drivers are supported:
+them to C header file that can be included in Wi-Fi shell module.
+
+Convert CA format from PEM to DER
+=======================================
```
```suggestion
==============================
```

Heading format should align till end of the text

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -29,6 +29,13 @@ LOG_MODULE_REGISTER(net_wifi_shell, LOG_LEVEL_INF);
 
 #include "net_shell_private.h"
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE
+#include "ca-cert.h"
+#include "client-cert.h"
+#include "client-key.h"
+#include "client-cert.h"
+#include "client-key.h"
```
Ah, you want to use the certs here in the shell. I think user should be able to manipulate the certs at runtime, there is tls-certificate shell for that, we can use it to let the user to manipulate the certificates.

MaochenWang1 replied:
Hi @jukkar  this is @krish2718 's suggestions that I should add the shell cmd to set the enterprise cert/key for wifi, you mean I should remove the shell cmd of enterprise_creds here?

krish2718 replied:
>  there is tls-certificate shell for that, we can use it to let the user to manipulate the certificates.

I had tried that, it's incredibly difficult to use, need to have custom configs to enable larger buffers and default SHELL RX ring is small, so, need to load `cred buf XXXX` multiple times at least 10 for a normal cert, we can certainly add this in the future, but just copying pem/header and rebuilding would be easy esp. as certs don't change often.

krish2718 replied:
Yes, the idea is to overwrite the cert/C header in `test_certs` and then rebuild, I can push a change that automates cert->C header generation instead of user running manual commands.

jukkar replied:
Even if we store the certs in samples or somewhere else, the bigger question is how to manage them. I mean we have the TLS credentials API for storing the certificates, so passing them around like this in this PR looks a bit problematic. If using the TLS cred API the certs could be stored in secure storage instead of RAM / flash which is not secure at all.

krish2718 replied:
I took at a stab auto-generating these cert headers: https://github.com/krish2718/zephyr/commit/7254cae3f28f861bec7d06d5f192e60b105daad9

krish2718 replied:
> If using the TLS cred API the certs could be stored in secure storage instead of RAM / flash which is not secure at all.

This is purely for testing and not for production.

krish2718 replied:
> Zephyr has a certificate storage support, see tls_credential API https://docs.zephyrproject.org/latest/doxygen/html/group__tls__credentials.html. So instead of directly passing certs to around, couldn't we utilize the cert API for this?

Yes, we can use that, but for non-secure storage, we need to write a sample and then call `tls_credential_add` or use TLS `cred` shell, while in the sample we can load certs at compile time, I was trying to get `shell` working at compile time as shell at runtime is too much trouble.

jukkar replied:
> I took at a stab auto-generating these cert headers: [krish2718@7254cae](https://github.com/krish2718/zephyr/commit/7254cae3f28f861bec7d06d5f192e60b105daad9)

Looks nice. Do you know if openssl always available in the build environment, for example in windows?

jukkar replied:
> This is purely for testing and not for production.

Understood. I just want to make it possible so that the user is able to change the certs "easily" if they are stored next to wifi-shell. Which means either pem or der format files.

krish2718 replied:
The convertion to DEF was done so that we can use `xxd` to generate a C header, MbedTLS PEM/DER is independent to this.

MaochenWang1 replied:
Hi @krish2718  thanks for your commit

> I took at a stab auto-generating these cert headers: [krish2718@7254cae](https://github.com/krish2718/zephyr/commit/7254cae3f28f861bec7d06d5f192e60b105daad9)

MaochenWang1 replied:
Hi @jukkar @krish2718 can this PR be approved, or what should I continue to change?

krish2718 replied:
>> I took at a stab auto-generating these cert headers: [krish2718@7254cae](https://github.com/krish2718/zephyr/commit/7254cae3f28f861bec7d06d5f192e60b105daad9)

> Looks nice. Do you know if openssl always available in the build environment, for example in windows?

Yes, OpenSSL is available for Windows and Mac.

jukkar replied:
As the certs are only used for testing, and it seems to be quite difficult to solve this otherwise, and to get this PR forward, I am reluctantly inclined to accept this. But I want the cert merging to be easier for the end user, there is no need for the user to create the header file as that can be done automatically (if using der files) using `generate_inc_file_for_target` CMake macro.

krish2718 replied:
> Hi @jukkar @krish2718 can this PR be approved, or what should I continue to change?

I will share a commit using `generate_inc_file_for_target` and some re-org and then you can squash that in to your commit in this PR.

krish2718 replied:
> generate_inc_file_for_target

This was a great suggestion, thanks. It actually can take PEM and convert C header (used in Zephyr already), so, did the following changes: https://github.com/krish2718/zephyr/commit/bd3ecb62cccc3a12b3b0d245a582281cce807951 

1. Moved certs to `samples/net/wifi/test_certs`
2. Modified `wifi_shell.c` and corresponding CMake file to auto-generate the headers.

This approach looks much better.

@MaochenWang1  Please cherry-pick my commit and `fixup` with original one, I have only build-tested this, so, please test to make sure that certs are actually working.

MaochenWang1 replied:
Hi @krish2718 thanks for your help, I cherry-pick your commit, and I use der to replace pem, as I found generate_inc_file_for_target only works for der to .h, please review @jukkar

MaochenWang1 replied:
there are compliance check warnings that I don't know how to fix:
![image](https://github.com/user-attachments/assets/52f16fe1-a896-4111-89c6-ddc76f1a5ea7)
but I see in /zephyr/samples/net/sockets/echo_client/src, there is echo-apps-cert.der

krish2718 replied:
I was finally able to get WPA2-EAP-TLS working on nRF7002DK, see my branch for latest fixes: https://github.com/krish2718/zephyr/tree/main_zephyr_10 I have squashed the commit, so, you can just cherry-pick and rebase and it's good to go. Apart from above, mainly added

* Heap adjustments
* SSL content length fix
* Add missing MbedTLS configs

I am working on adding individual configurations for Entperprise methods e.g., choose EAP-TLS only or EAP-PEAP only and we also need to move the respective MbedTLS dependencies to the individual methods, will submit a new PR for that.

krish2718 replied:
Also, https://github.com/krish2718/sdk-hostap/commit/61e37b46571ef25efdaef7744e46c5033ff30bab this fix is needed, I am not sure if this is applicable for all Enterprise methods or just EAP-TLS, I am sure there is a proper way to fix this, but MbedTLS itself hardcoded the length to `48` in the callback invocation see https://github.com/Mbed-TLS/mbedtls/blob/195e1647b24d231521675c0aab6d4715bb579be5/library/ssl_tls.c#L9114

MaochenWang1 replied:
Hi @jukkar  @krish2718  please review again.

MaochenWang1 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -88,6 +88,8 @@ enum net_request_wifi_cmd {
 	NET_REQUEST_WIFI_CMD_AP_STA_DISCONNECT,
```
thanks

MaochenWang1 commented on subsys/net/l2/wifi/test_certs/ca-cert.h:
```diff
@@ -0,0 +1,133 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+const unsigned char ca_cert_test[] = {
```
actually these are suggestions from Chaitanya

jukkar replied:
Yes, Chaitanya mentioned that separately. I just don't feel comfortable introducing hard coded certs in the core stack. Those are fine in samples and tests but introducing them in other places is problematic.

MaochenWang1 replied:
Hi @jukkar @krish2718 can you share the common solution that I can follow?

jukkar replied:
Personally I would add a sample that hosts these test certificates, they are just in wrong place when placed next to wifi shell.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1074,6 +1101,30 @@ static int cmd_wifi_ps_timeout(const struct shell *sh, size_t argc, char *argv[]
+
+	context.sh = sh;
+
+	params.ca_cert = (uint8_t *)ca_cert_test;
+	params.ca_cert_len = ca_cert_test_len;
+	params.client_cert = (uint8_t *)client_cert_test;
+	params.client_cert_len = client_cert_test_len;
+	params.client_key = (uint8_t *)client_key_test;
+	params.client_key_len = client_key_test_len;
```
Zephyr has a certificate storage support, see `tls_credential` API https://docs.zephyrproject.org/latest/doxygen/html/group__tls__credentials.html. So instead of directly passing certs to around, couldn't we utilize the cert API for this?

krish2718 replied:
We have the same discussion in three threads, let's use https://github.com/zephyrproject-rtos/zephyr/pull/74847#discussion_r1705361899 as the main one.

jukkar replied:
The proposed API using `struct wifi_enterprise_creds_params` looks like a public API to me and not for testing, or did I misunderstood this?

MaochenWang1 replied:
Hi @jukkar The proposed API using struct wifi_enterprise_creds_params, it's not only used for testing in wifi shell, also used by wifi driver to set the enterprise certs when connecting with enterprise mode AP

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -30,11 +30,20 @@ LOG_MODULE_REGISTER(net_wifi_shell, LOG_LEVEL_INF);
-#include "client-key.h"
+static const char ca_cert_test[] = {
+	#include <wifi_enterprise_test_certs/ca.der.inc>
+	'\0'
```
@krish2718 why do we need to add `\0` here, I do not see a need for that in echo-server sample for example.

krish2718 replied:
We can remove that, I added it based on `thrift`, but it's just specific to `thrift` (https://github.com/zephyrproject-rtos/zephyr/commit/218f787e2619851ca17a67780f144c62619421ff )

MaochenWang1 replied:
removed '\0'

jukkar reviewed (commented):
The cert auto-generation commit is a fixup commit that is fixing another commit in the same PR, we usually do not have those in zephyr so this needs to be squashed to the one that is introducing the certs.

krish2718 committed:
modules: hostap: Fix heap requirements for Enterprise

For Enterprise crypto MbedTLS needs more heap either separate pool or
libc heap, based on experiments 55000 was arrived for a successful
WPA2-EAP-TLS association.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
modules: hostap: Fix MbedTLS configs

For Wi-Fi Enterprise testing these configurations are necessary.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2484,12 +2559,14 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		  "[-m, --bssid]: MAC address of the AP (BSSID).\n"
 		  "[-t, --timeout]: Timeout for the connection attempt (in seconds).\n"
+		  "[-a, --anon-id]: Anonymous identity for enterprise mode.\n"
+		  "[-K, --key-passwd]: Private key passwd for enterprise mode.\n"
```
I tried giving wrong password for the private key but it still worked :), any ideas?

MaochenWang1 replied:
for enterprise mode, private_key_passwd is needed. but why you can work with wrong password, this can be debugged in the future.

krish2718 replied:
Sure, but please log a bug once this PR is merged to track.

MaochenWang1 replied:
okay

krish2718 commented:
Thanks a lot for your patience @MaochenWang1 , if you can cherry-pick fixes from my branches and address few remaining comments then we can merge it.

krish2718 commented on modules/hostap/src/supp_main.c:
```diff
@@ -61,12 +61,15 @@ static const struct wifi_mgmt_ops mgmt_ops = {
 #endif /* CONFIG_AP */
 	.dpp_dispatch = supplicant_dpp_dispatch,
+	.pmksa_flush = supplicant_pmksa_flush,
+	.enterprise_creds = supplicant_add_enterprise_creds,
```
need to be protected by `CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE`

MaochenWang1 replied:
added

MaochenWang1 commented on modules/hostap/src/supp_api.c:
```diff
@@ -321,6 +328,61 @@ static inline enum wifi_security_type wpas_key_mgmt_to_zephyr(int key_mgmt, int
+		goto out;
+	}
+
+	memcpy((void *)&enterprise_creds, (void *)creds,
```
I think it's not needed, as before every new connection, enterprise_creds will be set as new one.
and handle the disconnection case is kind of complex, as there is manually disconnection, and there is disconnection from AP or link lost, etc.

MaochenWang1 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2604,6 +2672,15 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
+	SHELL_CMD_ARG(enterprise_creds, NULL,
+		     "Set enterprise credentials.\n"
+		     "You can replace subsys/net/l2/wifi/test_certs with custom credentials.\n",
+		     cmd_wifi_set_enterprise_creds, 1, 0),
```
can we just keep this cmd to use shell to add credentials? I think we can remove this cmd when the full enterprise modes are upstreamed and verified.

krish2718 replied:
> can we just keep this cmd to use shell to add credentials? 

Any specific reason? we are doing it automatically anyway.

krish2718 replied:
I see that it's already removed.

MaochenWang1 commented on include/zephyr/net/wifi.h:
```diff
@@ -50,14 +50,14 @@ enum wifi_security_type {
 	/** WPA/WPA2/WPA3 PSK security. */
 	WIFI_SECURITY_TYPE_WPA_AUTO_PERSONAL,
+	/** EAP TLS security - Enterprise. */
+	WIFI_SECURITY_TYPE_EAP_TLS,
```
updated

krish2718 commented on include/zephyr/net/wifi.h:
```diff
@@ -50,14 +50,16 @@ enum wifi_security_type {
 	/** WPA/WPA2/WPA3 PSK security. */
 	WIFI_SECURITY_TYPE_WPA_AUTO_PERSONAL,
+	/** EAP security - Enterprise. */
+	WIFI_SECURITY_TYPE_EAP,
```
this would break shell as the enumerations change, so, better leave it in the original place.

MaochenWang1 replied:
actually in the feature, we will add more enterprise mode wifi_security_type, like:
	/** EAP PEAP MSCHAPV2 security - Enterprise. */
	WIFI_SECURITY_TYPE_EAP_PEAP_MSCHAPV2,
	/** EAP PEAP GTC security - Enterprise. */
	WIFI_SECURITY_TYPE_EAP_PEAP_GTC,
	/** EAP TTLS MSCHAPV2 security - Enterprise. */
	WIFI_SECURITY_TYPE_EAP_TTLS_MSCHAPV2,
	/** EAP PEAP TLS security - Enterprise. */
	WIFI_SECURITY_TYPE_EAP_PEAP_TLS,
	/** FT-PSK security */
	WIFI_SECURITY_TYPE_FT_PSK,
	/** FT-SAE security */
	WIFI_SECURITY_TYPE_FT_SAE,
	/** FT-EAP security */
	WIFI_SECURITY_TYPE_FT_EAP,
	/** FT-EAP-SHA384 security */
	WIFI_SECURITY_TYPE_FT_EAP_SHA384,
and the shell cmd like:
		  "[-k, --key-mgmt]: Key Management type (valid only for secure SSIDs)\n"
		  "0:None, 1:WPA2-PSK, 2:WPA2-PSK-256, 3:SAE, 4:WAPI, 5:WEP, 6: WPA-PSK\n"
		  "7: WPA-Auto-Personal, 8: DPP, 9: EAP-TLS, 10: EAP-PEAP-MSCHAPv2\n"
		  "11: EAP-PEAP-GTC, 12: EAP-TTLS-MSCHAPv2, 13: EAP-PEAP-TLS\n"

	so it's weird to keep the old one WIFI_SECURITY_TYPE_EAP. I suggest to directly remove WIFI_SECURITY_TYPE_EAP.
	if must keep it to keep the backwards, it should move to place with other enterprise security type.

krish2718 replied:
> actually in the feature, we will upstream more enterprise mode wifi_security_type(already support in our downstream), 

Sure, but we will remove EAP eventually, though this is just an Enumeration, still better to keep is compatible (See https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated for guidelines regards to API).

We can mark EAP as deprecated and recommend EAP_TLS as the replacement and in the future can remove EAP.

MaochenWang1 replied:
> this would break shell as the enumerations change
could you share which case the shell will be broken? As the shell cmd has prompt in console, user can check the cmd usage, and the EAP type is not widely used currently in zephyr.
> though this is just an Enumeration, still better to keep is compatible (See https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated for guidelines regards to API
From this page, I only see that mark "function declarations or macro definitions" as deprecated

krish2718 replied:
> > this would break shell as the enumerations change
> 
> could you share which case the shell will be broken? As the shell cmd has prompt in console, user can check the cmd usage, and the EAP type is not widely used currently in zephyr.

The numbers for the security types are changed, not just EAP, but for WEP and WPA-PSK too.

> > though this is just an Enumeration, still better to keep is compatible (See https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated for guidelines regards to API
> 
> From this page, I only see that mark "function declarations or macro definitions" as deprecated. How to mark the enumeration as deprecated?

As I said, no clear instructions for Enumeration, but if we don't need to break it, it's best both old and new co-exist, at least for a while.

MaochenWang1 replied:
udpated

krish2718 replied:
Thank you.

MaochenWang1 commented:
Hi @jukkar please review, thanks

MaochenWang1 commented:
Hi @jukkar  please review, this PR has pending too long, and faces conflicts several times

jukkar reviewed (approved):
Thank you for patience, LGTM.
